mypyc/irbuild/function.py

```diff
@@ -352,7 +369,7 @@ def gen_func_ir(builder: IRBuilder,
     else:
         assert isinstance(fn_info.fitem, FuncDef)
         func_decl = builder.mapper.func_to_decl[fn_info.fitem]
-        if fn_info.is_decorated:
+        if fn_info.is_decorated or is_singledispatch_main_func:
```

Comment:
Why do we do this?

Reply:
This check for whether the function is decorated is used to determine whether to preserve any name mangling applied to `fn_info.name` by creating a new `FuncDecl` instead of reusing the one in `func_to_decl`. If we don't do this, then both the dispatch function and the singledispatch main function end up getting the same name.

Previously, we didn't need this because this because the main singledispatch function happened to be decorated, but 3419f3918c1cb9f4350e520be735d087aa226ada removed that decorator.

I put more information in the commit message for 5c775644caec3f6ee50d7fbd96470a3b6f07aded.